File: src/OvertimeCalculator.java
import java.util.InputMismatchException;
import java.util.Scanner;

/**
 * Programa de consola para calcular el valor y total de horas extras de un empleado.
 *
 * <p>Solicita al usuario los datos necesarios (nombre, salario mensual, horas trabajadas
 * y horas ordinarias contractuales) y calcula el valor de la hora extra y el monto total
 * a pagar por dichas horas.</p>
 *
 * @author <a href="https://github.com/Elias-Santander">Elias Santander</a>
 * @version 1.0
 */
public class OvertimeCalculator {

    // -----------------------------------------------------------------------
    // Constantes de negocio
    // -----------------------------------------------------------------------

    /** Número de días laborables en un mes estándar. */
    private static final int WORKING_DAYS_PER_MONTH = 30;

    /** Número de horas laborables en un día estándar. */
    private static final int WORKING_HOURS_PER_DAY = 8;

    /** Factor de recargo aplicado sobre el valor de la hora ordinaria para calcular la hora extra. */
    private static final double OVERTIME_SURCHARGE_FACTOR = 1.5;

    // -----------------------------------------------------------------------
    // Punto de entrada
    // -----------------------------------------------------------------------

    /**
     * Método principal que ejecuta el flujo completo del programa.
     *
     * @param args argumentos de línea de comandos (no se utilizan)
     */
    public static void main(String[] args) {

        // Iniciamos el scanner para leer entrada del usuario
        try (Scanner scanner = new Scanner(System.in)) {

            printHeader();

            // --- Recolección de datos ---
            String employeeName   = readEmployeeName(scanner);
            double monthlySalary  = readPositiveDouble(scanner, "Ingrese el salario mensual bruto ($): ");
            double hoursWorked    = readPositiveDouble(scanner, "Ingrese el total de horas trabajadas este mes: ");
            double contractHours  = readContractHours(scanner);

            // --- Cálculo ---
            OvertimeResult result = calculateOvertime(monthlySalary, hoursWorked, contractHours);

            // --- Reporte ---
            printReport(employeeName, monthlySalary, hoursWorked, contractHours, result);

        } catch (Exception e) {
            // Captura cualquier excepción inesperada y muestra mensaje amigable
            System.err.println("\n[ERROR] Ocurrió un error inesperado: " + e.getMessage());
        }
    }

    // -----------------------------------------------------------------------
    // Métodos de lectura / validación
    // -----------------------------------------------------------------------

    /**
     * Imprime el encabezado decorativo del programa.
     */
    private static void printHeader() {
        System.out.println("=".repeat(55));
        System.out.println("       CALCULADORA DE HORAS EXTRAS");
        System.out.println("=".repeat(55));
    }

    /**
     * Lee y valida el nombre del empleado (no puede estar vacío).
     *
     * @param scanner instancia de {@link Scanner} para leer la entrada
     * @return nombre del empleado ingresado por el usuario
     */
    private static String readEmployeeName(Scanner scanner) {
        String name;
        while (true) {
            System.out.print("Ingrese el nombre del empleado: ");
            name = scanner.nextLine().trim();
            if (!name.isEmpty()) {
                return name;
            }
            // Validación: el nombre no puede estar vacío
            System.out.println("  [!] El nombre no puede estar vacío. Intente nuevamente.");
        }
    }

    /**
     * Lee un número decimal positivo desde la consola, reintentando si el valor
     * ingresado no es válido o es menor o igual a cero.
     *
     * @param scanner instancia de {@link Scanner} para leer la entrada
     * @param prompt  mensaje que se muestra al usuario antes de leer el valor
     * @return valor decimal positivo ingresado por el usuario
     */
    private static double readPositiveDouble(Scanner scanner, String prompt) {
        while (true) {
            System.out.print(prompt);
            try {
                double value = scanner.nextDouble();
                scanner.nextLine(); // consumir el salto de línea pendiente
                if (value > 0) {
                    return value;
                }
                // Validación: el valor debe ser mayor que cero
                System.out.println("  [!] El valor debe ser mayor que cero. Intente nuevamente.");
            } catch (InputMismatchException e) {
                // El usuario ingresó algo que no es un número
                System.out.println("  [!] Entrada inválida. Por favor ingrese un número válido.");
                scanner.nextLine(); // limpiar buffer para evitar bucle infinito
            }
        }
    }

    /**
     * Lee las horas contractuales ordinarias del mes, ofreciendo un valor
     * por defecto calculado a partir de {@code WORKING_DAYS_PER_MONTH} y
     * {@code WORKING_HOURS_PER_DAY}.
     *
     * @param scanner instancia de {@link Scanner} para leer la entrada
     * @return total de horas ordinarias pactadas en el contrato
     */
    private static double readContractHours(Scanner scanner) {
        // Calculamos el valor por defecto (ej: 30 días × 8 horas = 240)
        double defaultHours = (double) WORKING_DAYS_PER_MONTH * WORKING_HOURS_PER_DAY;
        System.out.printf("Ingrese las horas ordinarias contractuales del mes [%.0f]: ", defaultHours);

        String input = scanner.nextLine().trim();

        // Si el usuario presiona Enter sin ingresar nada, usamos el valor por defecto
        if (input.isEmpty()) {
            System.out.printf("  [i] Se usará el valor por defecto: %.0f horas.%n", defaultHours);
            return defaultHours;
        }

        // Si ingresó un valor, lo validamos
        try {
            double value = Double.parseDouble(input);
            if (value > 0) {
                return value;
            }
            System.out.println("  [!] Valor inválido. Se usará el valor por defecto.");
        } catch (NumberFormatException e) {
            // El usuario ingresó texto no numérico; usamos el valor por defecto
            System.out.println("  [!] Entrada inválida. Se usará el valor por defecto.");
        }
        return defaultHours;
    }

    // -----------------------------------------------------------------------
    // Lógica de negocio
    // -----------------------------------------------------------------------

    /**
     * Calcula el valor de la hora ordinaria, el valor de la hora extra y
     * el monto total a pagar por las horas extras trabajadas.
     *
     * <p>Fórmulas utilizadas:</p>
     * <ul>
     *   <li>Hora ordinaria  = salario mensual / horas contractuales</li>
     *   <li>Hora extra      = hora ordinaria × {@value #OVERTIME_SURCHARGE_FACTOR}</li>
     *   <li>Horas extra     = horas trabajadas − horas contractuales (mínimo 0)</li>
     *   <li>Total horas extra = horas extra × valor hora extra</li>
     * </ul>
     *
     * @param monthlySalary  salario mensual bruto del empleado
     * @param hoursWorked    total de horas efectivamente trabajadas en el mes
     * @param contractHours  horas ordinarias pactadas en el contrato para el mes
     * @return objeto {@link OvertimeResult} con todos los valores calculados
     */
    private static OvertimeResult calculateOvertime(double monthlySalary,
                                                    double hoursWorked,
                                                    double contractHours) {
        // Valor de la hora ordinaria
        double ordinaryHourlyRate = monthlySalary / contractHours;

        // Valor de la hora extra con recargo
        double overtimeHourlyRate = ordinaryHourlyRate * OVERTIME_SURCHARGE_FACTOR;

        // Horas extras (no puede ser negativo si el empleado no superó el límite)
        double overtimeHours = Math.max(0, hoursWorked - contractHours);

        // Monto total a pagar por horas extras
        double totalOvertimePay = overtimeHours * overtimeHourlyRate;

        return new OvertimeResult(ordinaryHourlyRate, overtimeHourlyRate, overtimeHours, totalOvertimePay);
    }

    // -----------------------------------------------------------------------
    // Reporte
    // -----------------------------------------------------------------------

    /**
     * Imprime el reporte detallado de horas extras en consola.
     *
     * @param employeeName   nombre del empleado
     * @param monthlySalary  salario mensual bruto
     * @param hoursWorked    horas trabajadas en el mes
     * @param contractHours  horas ordinarias contractuales
     * @param result         resultado del cálculo de horas extras
     */
    private static void printReport(String employeeName,
                                    double monthlySalary,
                                    double hoursWorked,
                                    double contractHours,
                                    OvertimeResult result) {
        System.out.println();
        System.out.println("=".repeat(55));
        System.out.println("           DETALLE DE HORAS EXTRAS");
        System.out.println("=".repeat(55));
        System.out.printf("  Empleado              : %s%n",          employeeName);
        System.out.printf("  Salario mensual       : $ %,.2f%n",     monthlySalary);
        System.out.printf("  Horas contractuales   : %.2f hrs%n",    contractHours);
        System.out.printf("  Horas trabajadas      : %.2f hrs%n",    hoursWorked);
        System.out.println("-".repeat(55));
        System.out.printf("  Valor hora ordinaria  : $ %,.2f%n",     result.ordinaryHourlyRate());
        System.out.printf("  Valor hora extra (%sx): $ %,.2f%n",
                           OVERTIME_SURCHARGE_FACTOR,                 result.overtimeHourlyRate());
        System.out.printf("  Total horas extras    : %.2f hrs%n",    result.overtimeHours());
        System.out.println("=".repeat(55));
        System.out.printf("  TOTAL A PAGAR (H.E.)  : $ %,.2f%n",    result.totalOvertimePay());
        System.out.println("=".repeat(55));

        // Mensaje informativo si no hubo horas extras
        if (result.overtimeHours() == 0) {
            System.out.println("  [i] El empleado no registró horas extras este mes.");
        }
        System.out.println();
    }

    // -----------------------------------------------------------------------
    // Clase de resultado (Record inmutable)
    // -----------------------------------------------------------------------

    /**
     * Contenedor inmutable con los resultados del cálculo de horas extras.
     *
     * @param ordinaryHourlyRate  valor de la hora ordinaria
     * @param overtimeHourlyRate  valor de la hora extra con recargo
     * @param overtimeHours       cantidad de horas extras trabajadas
     * @param totalOvertimePay    monto total a pagar por horas extras
     */
    private record OvertimeResult(
            double ordinaryHourlyRate,
            double overtimeHourlyRate,
            double overtimeHours,
            double totalOvertimePay) {
    }
}

File: src/OvertimeCalculator.cs
using System;

/// <summary>
/// Programa de consola para calcular el valor y total de horas extras de un empleado.
/// Solicita al usuario los datos necesarios (nombre, salario mensual, horas trabajadas
/// y horas ordinarias contractuales) y calcula el valor de la hora extra y el monto total
/// a pagar por dichas horas.
/// </summary>
/// <author>Elias Santander</author>
/// <version>1.0</version>
internal class OvertimeCalculator
{
    // -----------------------------------------------------------------------
    // Constantes de negocio
    // -----------------------------------------------------------------------

    /// <summary>Número de días laborables en un mes estándar.</summary>
    private const int WorkingDaysPerMonth = 30;

    /// <summary>Número de horas laborables en un día estándar.</summary>
    private const int WorkingHoursPerDay = 8;

    /// <summary>Factor de recargo aplicado sobre el valor de la hora ordinaria para calcular la hora extra.</summary>
    private const double OvertimeSurchargeFactor = 1.5;

    // -----------------------------------------------------------------------
    // Punto de entrada
    // -----------------------------------------------------------------------

    /// <summary>
    /// Método principal que ejecuta el flujo completo del programa.
    /// </summary>
    /// <param name="args">Argumentos de línea de comandos (no se utilizan).</param>
    private static void Main(string[] args)
    {
        try
        {
            PrintHeader();

            // --- Recolección de datos ---
            string employeeName  = ReadEmployeeName();
            double monthlySalary = ReadPositiveDouble("Ingrese el salario mensual bruto ($): ");
            double hoursWorked   = ReadPositiveDouble("Ingrese el total de horas trabajadas este mes: ");
            double contractHours = ReadContractHours();

            // --- Cálculo ---
            OvertimeResult result = CalculateOvertime(monthlySalary, hoursWorked, contractHours);

            // --- Reporte ---
            PrintReport(employeeName, monthlySalary, hoursWorked, contractHours, result);
        }
        catch (Exception ex)
        {
            // Captura cualquier excepción inesperada y muestra mensaje amigable
            Console.Error.WriteLine($"\n[ERROR] Ocurrió un error inesperado: {ex.Message}");
        }
    }

    // -----------------------------------------------------------------------
    // Métodos de lectura / validación
    // -----------------------------------------------------------------------

    /// <summary>
    /// Imprime el encabezado decorativo del programa.
    /// </summary>
    private static void PrintHeader()
    {
        Console.WriteLine(new string('=', 55));
        Console.WriteLine("       CALCULADORA DE HORAS EXTRAS");
        Console.WriteLine(new string('=', 55));
    }

    /// <summary>
    /// Lee y valida el nombre del empleado (no puede estar vacío).
    /// </summary>
    /// <returns>Nombre del empleado ingresado por el usuario.</returns>
    private static string ReadEmployeeName()
    {
        while (true)
        {
            Console.Write("Ingrese el nombre del empleado: ");
            string? input = Console.ReadLine()?.Trim();

            // Validación: el nombre no puede estar vacío ni nulo
            if (!string.IsNullOrEmpty(input))
                return input;

            Console.WriteLine("  [!] El nombre no puede estar vacío. Intente nuevamente.");
        }
    }

    /// <summary>
    /// Lee un número decimal positivo desde la consola, reintentando si el valor
    /// ingresado no es válido o es menor o igual a cero.
    /// </summary>
    /// <param name="prompt">Mensaje que se muestra al usuario antes de leer el valor.</param>
    /// <returns>Valor decimal positivo ingresado por el usuario.</returns>
    private static double ReadPositiveDouble(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            string? input = Console.ReadLine();

            // Intentamos parsear el valor; si falla o es negativo, volvemos a pedir
            if (double.TryParse(input, out double value) && value > 0)
                return value;

            Console.WriteLine("  [!] Entrada inválida. Ingrese un número mayor que cero.");
        }
    }

    /// <summary>
    /// Lee las horas contractuales ordinarias del mes, ofreciendo un valor
    /// por defecto calculado a partir de <see cref="WorkingDaysPerMonth"/> y
    /// <see cref="WorkingHoursPerDay"/>.
    /// </summary>
    /// <returns>Total de horas ordinarias pactadas en el contrato.</returns>
    private static double ReadContractHours()
    {
        // Calculamos el valor por defecto (ej: 30 días × 8 horas = 240)
        double defaultHours = WorkingDaysPerMonth * WorkingHoursPerDay;
        Console.Write($"Ingrese las horas ordinarias contractuales del mes [{defaultHours:F0}]: ");

        string? input = Console.ReadLine()?.Trim();

        // Si el usuario presiona Enter sin ingresar nada, usamos el valor por defecto
        if (string.IsNullOrEmpty(input))
        {
            Console.WriteLine($"  [i] Se usará el valor por defecto: {defaultHours:F0} horas.");
            return defaultHours;
        }

        // Si ingresó un valor, lo validamos
        if (double.TryParse(input, out double value) && value > 0)
            return value;

        // Entrada inválida: fallback al valor por defecto
        Console.WriteLine("  [!] Entrada inválida. Se usará el valor por defecto.");
        return defaultHours;
    }

    // -----------------------------------------------------------------------
    // Lógica de negocio
    // -----------------------------------------------------------------------

    /// <summary>
    /// Calcula el valor de la hora ordinaria, el valor de la hora extra y
    /// el monto total a pagar por las horas extras trabajadas.
    /// <para>Fórmulas utilizadas:</para>
    /// <list type="bullet">
    ///   <item>Hora ordinaria  = salario mensual / horas contractuales</item>
    ///   <item>Hora extra      = hora ordinaria × <see cref="OvertimeSurchargeFactor"/></item>
    ///   <item>Horas extra     = horas trabajadas − horas contractuales (mínimo 0)</item>
    ///   <item>Total horas extra = horas extra × valor hora extra</item>
    /// </list>
    /// </summary>
    /// <param name="monthlySalary">Salario mensual bruto del empleado.</param>
    /// <param name="hoursWorked">Total de horas efectivamente trabajadas en el mes.</param>
    /// <param name="contractHours">Horas ordinarias pactadas en el contrato para el mes.</param>
    /// <returns>Instancia de <see cref="OvertimeResult"/> con todos los valores calculados.</returns>
    private static OvertimeResult CalculateOvertime(double monthlySalary,
                                                    double hoursWorked,
                                                    double contractHours)
    {
        // Valor de la hora ordinaria
        double ordinaryHourlyRate = monthlySalary / contractHours;

        // Valor de la hora extra con recargo
        double overtimeHourlyRate = ordinaryHourlyRate * OvertimeSurchargeFactor;

        // Horas extras (no puede ser negativo si el empleado no superó el límite)
        double overtimeHours = Math.Max(0, hoursWorked - contractHours);

        // Monto total a pagar por horas extras
        double totalOvertimePay = overtimeHours * overtimeHourlyRate;

        return new OvertimeResult(ordinaryHourlyRate, overtimeHourlyRate, overtimeHours, totalOvertimePay);
    }

    // -----------------------------------------------------------------------
    // Reporte
    // -----------------------------------------------------------------------

    /// <summary>
    /// Imprime el reporte detallado de horas extras en consola.
    /// </summary>
    /// <param name="employeeName">Nombre del empleado.</param>
    /// <param name="monthlySalary">Salario mensual bruto.</param>
    /// <param name="hoursWorked">Horas trabajadas en el mes.</param>
    /// <param name="contractHours">Horas ordinarias contractuales.</param>
    /// <param name="result">Resultado del cálculo de horas extras.</param>
    private static void PrintReport(string employeeName,
                                    double monthlySalary,
                                    double hoursWorked,
                                    double contractHours,
                                    OvertimeResult result)
    {
        Console.WriteLine();
        Console.WriteLine(new string('=', 55));
        Console.WriteLine("           DETALLE DE HORAS EXTRAS");
        Console.WriteLine(new string('=', 55));
        Console.WriteLine($"  Empleado              : {employeeName}");
        Console.WriteLine($"  Salario mensual       : $ {monthlySalary:N2}");
        Console.WriteLine($"  Horas contractuales   : {contractHours:F2} hrs");
        Console.WriteLine($"  Horas trabajadas      : {hoursWorked:F2} hrs");
        Console.WriteLine(new string('-', 55));
        Console.WriteLine($"  Valor hora ordinaria  : $ {result.OrdinaryHourlyRate:N2}");
        Console.WriteLine($"  Valor hora extra ({OvertimeSurchargeFactor}x): $ {result.OvertimeHourlyRate:N2}");
        Console.WriteLine($"  Total horas extras    : {result.OvertimeHours:F2} hrs");
        Console.WriteLine(new string('=', 55));
        Console.WriteLine($"  TOTAL A PAGAR (H.E.)  : $ {result.TotalOvertimePay:N2}");
        Console.WriteLine(new string('=', 55));

        // Mensaje informativo si no hubo horas extras
        if (result.OvertimeHours == 0)
            Console.WriteLine("  [i] El empleado no registró horas extras este mes.");

        Console.WriteLine();
    }

    // -----------------------------------------------------------------------
    // Clase de resultado (Record inmutable)
    // -----------------------------------------------------------------------

    /// <summary>
    /// Contenedor inmutable con los resultados del cálculo de horas extras.
    /// </summary>
    /// <param name="OrdinaryHourlyRate">Valor de la hora ordinaria.</param>
    /// <param name="OvertimeHourlyRate">Valor de la hora extra con recargo.</param>
    /// <param name="OvertimeHours">Cantidad de horas extras trabajadas.</param>
    /// <param name="TotalOvertimePay">Monto total a pagar por horas extras.</param>
    private record OvertimeResult(
        double OrdinaryHourlyRate,
        double OvertimeHourlyRate,
        double OvertimeHours,
        double TotalOvertimePay);
}
